### PR TITLE
Enable C# separate compilation

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Changes
- Make the name of the C# FuncExtension class internal so it is unique per .NET assembly in a multi-assembly Dafny project.

### Tests
- Add a test for C# separate compilation, named `separate-compilation/usesTimesTwo.dfy`, that compiles to a producer and a consumer Dafny program to two separate .cs files, and then further compiles them with dotnet using two .csproj files.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyRuntime/dafnyRuntime.dfy [hunk 1]
```diff
@@ -1,5 +1,5 @@
 method HasTuples() {
-  var has0 := ();
+  var has0 := (0);
```
Comment (robin-aws) [inline Source/DafnyRuntime/dafnyRuntime.dfy:2]:
```suggestion
  var has0 := ();
```

:)

**Review by robin-aws — COMMENTED** [review]
Very nice! I'll keep adding things to this test case to expose more issues like the tuples and FuncExtensions.

Nit, but let's qualify the PR title/commit message a bit more, since this is C# only for now and we know it's only partial support.

File: Test/comp/separate-compilation/Inputs/producer/TimesTwo.csproj [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net6.0</TargetFramework>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <Reference Include="DafnyRuntime">
+            <HintPath>..\..\..\..\..\Binaries\DafnyRuntime.dll</HintPath>
+            <HintPath>..\..\..\..\..\..\..\DafnyRuntime.dll</HintPath>
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Test/comp/separate-compilation/Inputs/producer/TimesTwo.csproj:11]: For running both through regular old lit and through the XUnit testrunner.

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Test/comp/separate-compilation/Inputs/producer/TimesTwo.csproj:11]: For running both through regular old lit and through the XUnit testrunner.

[the remark this answers is not in the document]
Comment (robin-aws) [reply Test/comp/separate-compilation/Inputs/producer/TimesTwo.csproj:11]: Yeesh - okay fair enough.

**Review by robin-aws — APPROVED** [review]

**Pull request merged**